---
title: "Price_ProblemSet9"
author: "Andrew Price"
date: "November 4, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rafalib)
knitr::opts_chunk$set(error = TRUE)

```

###Batch Effects

###Confounding Exercises

###1. Let’s compute the proportion of men who were accepted: What is the proportion of women that were accepted? 


```{r}
install.packages("ps")  
library(devtools)
install_github("genomicsclass/dagdata")
library(dagdata)
data(admissions)
print(admissions)

index = which(admissions$Gender==1) 
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
applied = sum(admissions$Number[index]) 
accepted/applied

index = which(admissions$Gender==0) 
accepted= sum(admissions$Number[index] * admissions$Percent[index]/100) 
applied = sum(admissions$Number[index]) 
accepted/applied
```

The proportion of women accepted was .303

###2.  Now that we have observed different acceptance rates between genders, test for the significance of this result. If you perform an independence test, what is the p-value? How can this be? This is referred to as Simpson’s Paradox. In the following questions we will try to decipher why this is happening.

```{r Question 2}
atab<-as.data.frame(matrix(0,2,2))
rownames(atab)<-c("Men","Women")
colnames(atab)<-c("Accepted","Rejected")
index = admissions$Gender==1
men=admissions[index,]
women=admissions[!index,]
atab[1,1]= sum(men$Number * men$Percent/100)
atab[1,2]= sum(men$Number*(1-men$Percent/100))
windex = which(admissions$Gender==0)
atab[2,1]= sum(women$Number * women$Percent/100)
atab[2,2]= sum(women$Number * (1-women$Percent/100))
atab
chisq.test(atab)$p.value

index = admissions$Gender==1
men = admissions[index,]
women = admissions[!index,]
print( data.frame( major=admissions[1:6,1],men=men[,3], women=women[,3]) )
```

The reason that this happens is that the data is confounded on another variable.

###3. We can quantify how “hard” a major is by using the percent of students that were accepted. Compute the percent that were accepted (regardless of gender) to each major and call this vector H. Which is the hardest major? 

```{r Question 3}
major = admissions[1:6,1]
men = admissions[1:6,]
women =admissions[7:12,]
H = (men$Number*men$Percent/100 + women$Number*women$Percent/100) / (men$Number+women$Number)
H
major[which.min(H)]

```

F is the hardest major

###4. What proportion is accepted for this major?

```{r Question 4}
min(H)
```

###5. For men, what is the correlation between the number of applications across majors and H? 

```{r Question 5}
cor(H,men$Number)
```

###6. For women, what is the correlation between the number of applications across majors and H? 

```{r Question 6}
cor(H,women$Number)
```

###7.Given the answers to the above, which best explains the differences in admission percentages when we combine majors? • A) We made a coding mistake when computing the overall admissions percentages. • B)There were more total number of women applications which made the denominator much bigger. • C) There is confounding between gender and preference for “hard” majors: females are more likely to apply to harder majors. • D)The sample size for the individual majors was not large enough to draw the correct conclusion.

Answer: C

###Discovering Batch Effects Exercises

###1.Familiarize yourself with the sampleInfo table. Note that some samples were processed at different times. This is an extraneous variable and should not affect the values in geneExpression. However, as we have seen in previous analyses, it does appear to have an effect so we will explore this here.

```{r Question 8}

library(rafalib)
library(Biobase)
library(GSE5859) ##Available from GitHub

data(GSE5859)
geneExpression = exprs(e)
sampleInfo = pData(e)
year = format(sampleInfo$date,"%y")
table(year,sampleInfo$ethnicity)

```

###2. For how many of these years do we have more than one ethnicity represented?

```{r Question 9}
tab=table(year,sampleInfo$ethnicity)
print(tab)
x=rowSums(tab!=0)
sum(x>=2)
```

2 of the years, 05 and 06

###3. Repeat the above exercise, but now,instead of year,consider the month as well.Specifically, instead of the year variable defined above use: For what proportion of these month.year values do we have more than one ethnicity represented?

```{r Question 10}
month.year = format(sampleInfo$date,"%m%y")
mytab=table(month.year,sampleInfo$ethnicity)
print(mytab)
myx=rowSums(mytab!=0)
mean(myx>=2)
```

.0476

###4. Perform a t-test (use rowttests) comparing CEU samples processed in 2002 to those processed in 2003. Then use the qvalue package to obtain q-values for each gene. How many genes have q-values < 0.05 ? 

```{r Question 11}

library(qvalue)
library(genefilter)
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("02","03") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

###5. What is the estimate of pi0 provided by qvalue:

```{r Question 12}
qval$pi0
```

###6. Now perform a t-test (use rowttests) comparing CEU samples processed in 2003 to those processed in 2004. Then use the q value package to obtain q-values for each gene.How many genes have q-values less than 0.05? 

```{r Question 13}
year = factor( format(sampleInfo$date,"%y") )
index = which(year%in% c("03","04") & sampleInfo$ethnicity=="CEU")
year = droplevels(year[index])
pval = rowttests(geneExpression[ ,index], year)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

###7. How many genes have q-values < 0.05 ?

```{r Question 14}
ethnicity = factor( sampleInfo$ethnicity)
index = which(ethnicity%in% c("CEU","ASN"))
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```

###8. How many genes have q-values < 0.05 ?

```{r Question 15}
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
index = which(ethnicity%in% c("CEU","ASN") & year=="05")
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```


###9. How many genes have q-values < 0.05 

```{r Question 16}
ethnicity = factor( sampleInfo$ethnicity)
year = factor( format(sampleInfo$date,"%y") )
indexASN = which(ethnicity%in% c("ASN") & year=="05")
indexCEU = which(ethnicity%in% c("CEU") & year=="02")
set.seed(3)
indexCEU<-sample(indexCEU,3)
index<-c(indexASN,indexCEU)
ethnicity = droplevels(ethnicity[index])
pval = rowttests(geneExpression[ ,index], ethnicity)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.05)
```


###Estimate with Linear Modeling Exercises

###1. Using the functions rowttests and qvalue compare the two groups.Because this is a smaller dataset which decreases our power, we will use the more lenient FDR cut-off of 10%. How many gene have q-values less than 0.1?


```{r Question 17}
library(GSE5859Subset) 
data(GSE5859Subset)
sex = sampleInfo$group
month = factor( format(sampleInfo$date,"%m")) 
table( sampleInfo$group, month)

sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
sum(qval$qvalue < 0.1)
```

###2. Note that sampleInfo$group here represents males and females. Thus we expect differences to be on chrY and, for genes that escape inactivation, chrX. Note that we do not expect many autosomal genes to be different between males and females. This gives us an opportunity to evaluate false and true positives with experimental data. For example, we evaluate results using the proportion genes of the list that are on chrX or chrY.For the list of genes with q<0.1 calculated in Modeling Batch Effects Exercises #1, what proportion of genes are on chrX or chrY?

```{r Question 18}
sex = factor(sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

###3. We can also check how many of the chromosomes X and Y genes we detected as different. How many are on Y?

```{r Question 19}
sex = factor(sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
(index = geneAnnotation$CHR[qvals<0.1]%in%c("chrY"))
```

8 genes are on Y chromosomes

###4. Now for the autosomal genes (not on chrX and chrY) for which q-value < 0.1 perform a t-test comparing samples processed in June to those processed in October. What proportion of these have p-values < 0.05?

```{r Question 20}
sex = factor( sex)
pval = rowttests(geneExpression, sex)$p.value
qval = qvalue(pval)
qvals = qval$qvalues
index = which(qvals<0.1 & !geneAnnotation$CHR%in%c("chrX","chrY"))
month = factor( format(sampleInfo$date,"%m"))
pval = rowttests(geneExpression[index,], month)$p.value
mean(pval<0.05)
```

###5.  The above result shows that the great majority of the autosomal genes show differences due to processing data. This provides further evidence that confounding is resulting in false positives. So we are going to try to model the month effect to better estimate the sex effect. We are going to use a linear model: Which of the following creates the appropriate design matrix?

```{r Question 21}
 (X = model.matrix(∼sex+month))
```

Answer) D

###6. How many of the q-values for the group comparison are now <0.1?

```{r Question 22}
X = model.matrix(~sex+month) 
i = 234 
y = geneExpression[i,] 
fit = lm(y~X) 
summary(fit)$coef

pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[2,4]} ) )

qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

17

###7. With this new list, what proportion of these are chrX and chrY

```{r Question 23}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```

###8. How many on Y or X?

```{r Question 24}
(index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY"))
```

20

###9. How many of the q-values for the month comparison are now <0.1?

```{r Question 25}
X = model.matrix(~sex+month)
pvals <- t( sapply(1:nrow(geneExpression),function(j){
  y <- geneExpression[j,]
  fit <- lm(y~X-1)
  summary(fit)$coef[3,4]} ) )

qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

###Factor Analysi Exercises

###1. Suppose you want to make an MA plot of the first two samples y = geneExpression[,1:2]. Which of the following projections gives us the projection of y so that column2 versus column 1 is an MA plot? A. y(1/√2 1/√2 1√2 −1/√2) B. y(1 1 1 −1) C. (1 1 1 −1)y D. (1 1 1 −1)y⊤ 

###2. Say Y is M ×N, in the SVD Y = UDV⊤ which of the following is not correct? • A) DV⊤ are the new coordinates for the projection U⊤Y • B) UD are the new coordinates for the projection Y V • C) D are the coordinates of the projection U⊤Y • D) U⊤Y is a projection from an N-dimensional to M-dimensional subspace.

###3. Based on these plots, which of the following you would say is true?

```{r Question 28}
y = geneExpression - rowMeans(geneExpression)
library(RColorBrewer)
mypar(1,2)
n=ncol(y)
cors=cor(y)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
o <- order(sampleInfo$date)
yo<-y[,o]
cors=cor(yo)
cols=colorRampPalette(rev(brewer.pal(11,"RdBu")))(100)
image(1:n,1:n,cors,xaxt="n",yaxt="n",col=cols,xlab="",ylab="",zlim=c(-1,1))
```

D) The fact that in the plot ordered by month we see two groups mainly driven by month,and within these we see subgroups driven by date,seems to suggest date more than month per se are the hidden factors.

###4. Based on the correlation plots above, we could argue that there are at least two hidden factors. Using PCA estimate these two factors. Specifically, apply the svd to y and use the first two PCs as estimates.

```{r Question 29}
s <- svd(y)
What <- s$v[,1:2] 
rownames(What)<-colnames(y)
round(What,2)
fit = s$u[,1:2]%*% (s$d[1:2]*t(What))
var(as.vector(fit))/var(as.vector(y))

( pcs = svd(y)$v[,1:2] )
```

B

###5. Plot each of the estimated factor ordered by date. Use color to denote month. The first factor is clearly related to date. Which of the following appear to be most different according to this factor?

```{r Question 30} 
month=format(sampleInfo$date, "%m")
pcs = svd(y)$v[,1:2]
o = order(sampleInfo$date)
cols = as.numeric(month)[o]
mypar(2,1)
for(i in 1:2){
  plot(pcs[o,i],col=cols,xaxt="n",xlab="")
  label = gsub("2005-","",sampleInfo$date[o])
  axis(1,1:ncol(y),label,las=2)
}
```

A) 23 June and 27 June

###6. How many PCs explain more than 10% of the variability? 

```{r Question 31}
s = svd(y)
varexplained = s$d^2/ sum(s$d^2)
plot(varexplained)
sum(varexplained>0.10)
```

2

###7. Which PC most correlates (negative or positive correlation) with month? 

```{r Question 32}
s = svd(y)
month = factor( format(sampleInfo$date,"%m"))
cors = cor( as.numeric(month),s$v)
plot(t(cors))
which.max(abs(cors))
```

The first PC

###8. What is this correlation (in absolute value)

```{r Question 33}
max(abs(cors))
```

###9. Which PC most correlates (negative or positive correlation) with sex?

```{r Question 34}
s = svd(y)
sex = sampleInfo$group
cors = cor( as.numeric(sex),s$v)
plot(t(cors))
which.max(abs(cors))
```

The first PC

###10. What is this correlation (in absolute value)? 

```{r Question 35}
max(abs(cors))
```

###11. Now instead of using month, which we have shown does not quite describe the batch, add the two estimated factors s$v[,1:2] to the linear model we used above. Apply this model to each gene and compute q-values for the sex difference. How many q-values < 0.1 for the sex comparison?

```{r Question 36}
X <- model.matrix(~sex+s$v[,1:2])
qval = qvalue(pvals)
qvals = qval$qvalues
sum(qvals<0.1)
```

###12. What proportion of the genes are on chromosomes X and Y

```{r Question 37}
index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY")
mean(index)
```


###Modeling Batch Effects with Factor Analysis Exercises

###1. How many genes have q-value < 0.1? FIX

```{r Question 38}
library(sva)
library(Biobase)
library(GSE5859Subset)
data(GSE5859Subset)
s <- svd(geneExpression-rowMeans(geneExpression))
cor(sampleInfo$group,s$v[,1])
sex = sampleInfo$group 
mod = model.matrix(~sex) 
svafit = sva(geneExpression,mod) 
head(svafit$sv)
for(i in 1:ncol(svafit$sv)){ 
  print( cor(s$v[,i],svafit$sv[,i]) ) }
svaX<-model.matrix(~sex+svafit$sv)

library(limma)
fit <- lmFit(geneExpression,svaX)
k <- 2
ses <- fit$stdev.unscaled[,k]*fit$sigma
ttest <- fit$coef[,k]/ses
pvals <- 2*pt(-abs(ttest),fit$df)
qval = qvalue(pvals)

qvals = qval$qvalues

sum(qvals<0.1)

```

13 genes

###2. How many of these genes are from chrY or chrX?

```{r Question 39}
(index = geneAnnotation$CHR[qvals<0.1]%in%c("chrX","chrY"))

```

12 genes
